src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.successfactors.source.input;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This {@code SuccessFactorsPartitionBuilder} will prepare the list of optimized splits containing start & end indices
+ * for each split including the optimized package size.
+ * <p>
+ * Default Split count is 8
+ * Max allowed Split count is 10
+ * Max allowed Package size is 1000
+ * <p>
+ * If the total fetch record count is less then equal to 2500 then only 1 split will be created.
+ * <p>
+ * Note: DEFAULT & MAX on split count and package size is for the Customer Preview.
+ */
+public class SuccessFactorsPartitionBuilder {
+  public static final int DEFAULT_SPLIT_COUNT = 8;
+  public static final long MAX_ALLOWED_PACKAGE_SIZE = 1000L;
+  private static final Logger LOG = LoggerFactory.getLogger(SuccessFactorsPartitionBuilder.class);
+
+  /**
+   * builds the list of {@code SuccessFactorsInputSplit}
+   *
+   * @param availableRecordCount available row count
+   * @param fetchRowCount        plugin property, number of rows to extract. 0 (zero) means the total record to fetch
+   *                             will be same as total number of available records.
+   * @return list of {@code SuccessFactorsInputSplit}
+   */
+  public List<SuccessFactorsInputSplit> buildSplits(long availableRecordCount, long fetchRowCount) {
+
+    List<SuccessFactorsInputSplit> list = new ArrayList<>();
+
+    long recordReadStartIndex = 1;
+    long actualRecordToExtract = (fetchRowCount == 0 ? availableRecordCount : fetchRowCount);
+    long recordReadEndIndex = Math.min(actualRecordToExtract, availableRecordCount);
+    if (actualRecordToExtract > availableRecordCount) {
+      actualRecordToExtract = availableRecordCount;
+      if (actualRecordToExtract <= 0) {
+        String msg = "As per the provided configuration no records were found for extraction.";
+        throw new IllegalArgumentException(msg);
+      }
+    }
+
+    // setting up the optimal package size values
+    long packageSize = Math.min(actualRecordToExtract, MAX_ALLOWED_PACKAGE_SIZE);
+
+    int splitCount = DEFAULT_SPLIT_COUNT;
+
+    // in case total fetch record count is less than the DEFAULT_PACKAGE_SIZE then Split count will be defaulted to 1.
+    if (actualRecordToExtract <= MAX_ALLOWED_PACKAGE_SIZE) {
```

Comment:
It looks like the number of splits is set to 1 or 8 depending on whether number of records <= 1000, is that correct?
That doesn't seem like an optimal approach. 
Can we do something like this: pick some optimalLoadOnSplit based on your testing. 
splitCount = actualRecordToExtract / optimalLoadOnSplit + (actualRecordToExtract % optimalLoadOnSplit != 0 ? 1 : 0)
then assign upto optimalLoadOnSplit records to each split. The last split might have less load on it.
See https://github.com/data-integrations/servicenow-plugins/blob/b957dd9c83885b96fc05087654af2a42628b2837/src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java#L116 for example

Reply:
Yes.
Number of splits = 1, if number of records <= 1000
Number of splits = 8, if number of records > 1000

Working on refactoring the `buildSplits()` method

Reply:
I've refactored the `buildSplits()` method. Please check and let me know if the approach is optimal or not.
4fe92b6e5245c37aa932504a41fe46312eb5d8ff